File: C++/find-in-mountain-array.cpp
// Time:  O(logn)
// Space: O(1)

/**
 * // This is the MountainArray's API interface.
 * // You should not implement it, or speculate about its implementation
 * class MountainArray {
 *   public:
 *     int get(int index);
 *     int length();
 * };
 */
class Solution {
public:
    int findInMountainArray(int target, MountainArray &mountainArr) {
        int peak = binarySearch(mountainArr, 0, mountainArr.length()-1,
                                [&](int x) { return mountainArr.get(x) >= mountainArr.get(x + 1); });
        int left = binarySearch(mountainArr, 0, peak,
                             [&](int x) { return mountainArr.get(x) >= target; });
        if (left <= peak && mountainArr.get(left) == target) {
            return left;
        }
        int right = binarySearch(mountainArr, peak, mountainArr.length() - 1,
                                 [&](int x) { return mountainArr.get(x) <= target; });
        if (right <= mountainArr.length() - 1 && mountainArr.get(right) == target) {
            return right;
        }
        return -1;
    }
    
private:
    int binarySearch(MountainArray &A, int left, int right,
                     const function<bool(int)>& check) {
        while (left <= right) {
            const auto& mid = left + (right - left) / 2;
            if (check(mid)) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
};

File: Python/find-in-mountain-array.py
# Time:  O(logn)
# Space: O(1)

# """
# This is MountainArray's API interface.
# You should not implement it, or speculate about its implementation
# """
class MountainArray(object):
   def get(self, index):
       """
       :type index: int
       :rtype int
       """
       pass

   def length(self):
       """
       :rtype int
       """
       pass


class Solution(object):
    def findInMountainArray(self, target, mountain_arr):
        """
        :type target: integer
        :type mountain_arr: MountainArray
        :rtype: integer
        """
        def binarySearch(A, left, right, check):
            while left <= right:
                mid = left + (right-left)//2
                if check(mid):
                    right = mid-1
                else:
                    left = mid+1
            return left
        
        peak = binarySearch(mountain_arr, 0, mountain_arr.length()-1,
                            lambda x: mountain_arr.get(x) >= mountain_arr.get(x+1))
        left = binarySearch(mountain_arr, 0, peak,
                            lambda x: mountain_arr.get(x) >= target)
        if left <= peak and mountain_arr.get(left) == target:
            return left
        right = binarySearch(mountain_arr, peak, mountain_arr.length()-1,
                             lambda x: mountain_arr.get(x) <= target)
        if right <= mountain_arr.length()-1 and mountain_arr.get(right) == target:
            return right
        return -1
